pkg/importer/data_processor.go

```diff
@@ -0,0 +1,217 @@
+	Convert ProcessingPhase = "Convert"
+	// Resize the disk image, this is only needed when the target contains a file system (block device do not need a resize)
+	Resize ProcessingPhase = "Resize"
+	// Complete is the phase where the entire process completed successfully and we can exit gracefully.
+	Complete ProcessingPhase = "Complete"
+	// Error is the phase in which we encountered an error and need to exit ungracefully.
+	Error ProcessingPhase = "Error"
+)
+
+// ErrRequiresScratchSpace indicates that we require scratch space.
+var ErrRequiresScratchSpace = fmt.Errorf("Scratch space required and none found")
+
+// ErrInvalidPath indicates that the path is invalid.
+var ErrInvalidPath = fmt.Errorf("Invalid transfer path")
+
+// DataSourceInterface is the interface all data providers should implement.
+type DataSourceInterface interface {
+	// Info() is called to get initial information about the data.
+	Info() (ProcessingPhase, error)
+	// Transfer() is called to transfer the data from the source to a temporary location.
+	Transfer(path string) (ProcessingPhase, error)
+	// Process() is called to do any special processing before giving the url to the data back to the processor
+	Process() (ProcessingPhase, error)
+	// Geturl returns the url that the data processor can use when converting the data.
+	GetURL() *url.URL
+	// Close closes any readers or other open resources.
+	Close() error
+}
+
+// DataProcessor holds the fields needed to process data from a data provider.
+type DataProcessor struct {
+	// currentPhase is the phase the processing is in currently.
+	currentPhase ProcessingPhase
+	// provider provides the data for processing.
+	source DataSourceInterface
+	// destination file. will be DataDir/disk.img if file system, or a block device (if a block device, then DataDir will not exist).
+	dataFile string
+	// dataDir path to target directory if it contains a file system.
+	dataDir string
+	// scratchDataDir path to the scratch space.
+	scratchDataDir string
+	// requestImageSize is the size we want the resulting image to be.
+	requestImageSize string
+	// available space is the available space before downloading the image
+	availableSpace int64
+}
+
+// NewDataProcessor create a new instance of a data processor using the passed in data provider.
+func NewDataProcessor(dataSource DataSourceInterface, dataFile, dataDir, scratchDataDir, requestImageSize string) *DataProcessor {
+	dp := &DataProcessor{
+		currentPhase:     Info,
+		source:           dataSource,
+		dataFile:         dataFile,
+		dataDir:          dataDir,
+		scratchDataDir:   scratchDataDir,
+		requestImageSize: requestImageSize,
+	}
+	// Calculate available space before doing anything.
+	dp.availableSpace = dp.calculateTargetSize()
+	return dp
+}
+
+// ProcessData is the main processing loop.
+func (dp *DataProcessor) ProcessData() error {
+	var err error
+	if util.GetAvailableSpace(dp.scratchDataDir) > int64(0) {
+		defer CleanDir(dp.scratchDataDir)
```

Comment:
Great that we are cleaning up, but why not do a preemptive clean as well?  An earlier one could have failed.

Reply:
Sure